### PR TITLE
Fixes for chapters 11 & 12

### PR DESCRIPTION
**Pull request by misicnenad**
This change is [Reviewable](https://reviewable.io/reviews/cairo-book/cairo-book/823)

### CODE REVIEW

File: src/ch11-05-hash.md [hunk 1]
```diff
@@ -36,15 +36,7 @@ To use hashes in your code, you must first import the relevant traits and functi
...
 ```rust
-{{#include ../listings/ch11-advanced-features/no_listing_04_hash_poseidon/src/lib.cairo:import}}
-
-{{#include ../listings/ch11-advanced-features/no_listing_04_hash_poseidon/src/lib.cairo:all}}
-```
-
-As our struct derives the `Hash` trait, we can call the function as follows for Poseidon hashing :
-
-```rust
-{{#rustdoc_include ../listings/ch11-advanced-features/no_listing_04_hash_poseidon/src/lib.cairo:main}}
```
Comment (misicnenad) [inline src/ch11-05-hash.md:47]: Duplicate of above code.

File: listings/ch11-advanced-features/no_listing_04_hash_poseidon/src/lib.cairo [hunk 2]
```diff
@@ -12,14 +9,9 @@ struct StructForHash {
     last: bool,
 }
 
-//ANCHOR: main
```
Comment (misicnenad) [inline listings/ch11-advanced-features/no_listing_04_hash_poseidon/src/lib.cairo:15]: Can just show a single code block.

Before:
image

After:
image

Related comment https://github.com/cairo-book/cairo-book/pull/823/files#r1620527665

File: src/appendix-03-derivable-traits.md [hunk 3]
```diff
@@ -53,23 +53,23 @@ The `Debug` trait is required, for example, when using the `assert_xx!` macros i
...
+The `Default` trait allows creation of a default value of a type. The most common default value is zero. All primitive types in the standard library implement `Default`.
 
-If you want to derive `Default` on a composite type, each of its elements must already implement `Default`. If you have an [`enum` type](ch06-01-enums.md), you must declare its default value by using the `#[default]` attribute on one of its variants.
+If you want to derive `Default` on a composite type, each of its elements must already implement `Default`. If you have an [`enum`](ch06-01-enums.md) type, you must declare its default value by using the `#[default]` attribute on one of its variants.
 
 An example:
 
 ```rust
-{{#include ../listings/appendinx/listing_07_default/src/lib.cairo}}
+{{#include ../listings/appendix/listing_07_default/src/lib.cairo}}
```
Comment (misicnenad) [inline src/appendix-03-derivable-traits.md:63]: Typo, causes no code to be displayed.
image

File: src/appendix-04-common-types-and-traits-and-cairo-prelude.md [hunk 4]
```diff
@@ -31,23 +31,23 @@ You can choose which prelude to use. For example, adding `edition = "2023_11"` i
...
-| `get_caller_address`      | `starknet::get_caller_address`                        | `get_caller_address()` is a function that returns the address of the caller of the contract. It can be used to identify the caller of a contract function.                             |
-| `get_contract_address`    | `starknet::info::get_contract_address`                | `get_contract_address()` is a function that returns the address of the current contract. It can be used to obtain the address of the contract being executed.                          |
+| Import                    | Path                                                  | Usage                                                                                                                                                                                     |
+| ------------------------- | ----------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
```
Comment (misicnenad) [inline src/appendix-04-common-types-and-traits-and-cairo-prelude.md:41]: Just formatted the file.

File: src/ch11-01-custom-data-structures.md [hunk 5]
```diff
@@ -21,10 +21,10 @@ Fortunately, Cairo provides a handy built-in [dictionary type](./ch03-02-diction
 simulate the behavior of mutable data structures. Let's first explore how to create a struct that contains, among others, a `Felt252Dict<T>`.
 
 > Note: Several concepts used in this chapter were already presented earlier in the book. We recommend checking out the following chapters if you need to revise them:
-[Structs](ch05-00-using-structs-to-structure-related-data.md),
-[Methods](./ch05-03-method-syntax.md),
-[Generic types](./ch08-00-generic-types-and-traits.md),
-[Traits](./ch08-02-traits-in-cairo.md).
+> [Structs](ch05-00-using-structs-to-structure-related-data.md),
```
Comment (misicnenad) [inline src/ch11-01-custom-data-structures.md:24]: Formatter for *.md* files changed it to this, will revert if necessary.

The UI is the same though

File: src/ch11-01-custom-data-structures.md [hunk 6]
```diff
@@ -109,6 +109,12 @@ pointer to allow using any type `T` in our data structure, as explained in the
 {{#include ../listings/ch11-advanced-features/no_listing_13_cust_struct_vect/src/lib.cairo:struct}}
 ```
 
+Since we again have `Felt252Dict<T>` as a struct member, we need to implement the `Destruct<T>` trait to tell the compiler how to make `NullableVec<T>` go out of scope.
+
+```rust,noplayground
+{{#include ../listings/ch11-advanced-features/no_listing_13_cust_struct_vect/src/lib.cairo:destruct}}
```
Comment (misicnenad) [inline src/ch11-01-custom-data-structures.md:115]: Missing piece of code showing that `Destruct` has to implemented too.

File: listings/ch11-advanced-features/no_listing_13_cust_struct_vect/src/lib.cairo [hunk 7]
```diff
@@ -24,11 +24,13 @@ struct NullableVec<T> {
 }
 // ANCHOR_END: struct
 
+// ANCHOR: destruct
```
Comment (misicnenad) [inline listings/ch11-advanced-features/no_listing_13_cust_struct_vect/src/lib.cairo:27]: Needed for https://github.com/cairo-book/cairo-book/pull/823/files#r1620536985

File: src/ch11-09-printing.md [hunk 8]
```diff
@@ -22,30 +22,30 @@ Here are some examples:
...
+The `println!` macro can do many kinds of formatting, and by default, the curly brackets tell `println!` to use formatting known as `Display` - output intended for direct end user consumption. The primitive types we’ve seen so far implement `Display` by default because there’s only one way you’d want to show a `1` or any other primitive type to a user. But with structs, the way `println!` should format the output is less clear because there are more display possibilities: Do we want commas or not? Do we want to print the curly brackets? Should all the fields be shown? Due to this ambiguity, Cairo doesn’t try to guess what we want, and structs don’t have a provided implementation of `Display` to use with `println!` and the `{}` placeholder.
```
Comment (misicnenad) [inline src/ch11-09-printing.md:48]: `you` -> `we` is an alignment with the rest of the text, where we usually talk in "we".

File: src/ch11-09-printing.md [hunk 9]
```diff
@@ -84,16 +84,16 @@ It is also possible to implement the `Display` trait for the `Point` struct usin
...
+This trait is very useful and is implemented by default for basic data types. It can also be simply derived for complex data types using the `#[derive(Debug)]` attribute, as long as all types they contain implement it. This eliminates the need to manually implement extra code to print complex data types.
 
- Note that `assert_xx!` macros used in tests require the provided values to implement the `Debug` trait, as they also print the result in case of assertion failure.
+Note that `assert_xx!` macros used in tests require the provided values to implement the `Debug` trait, as they also print the result in case of assertion failure.
 
-Please refer to the [Derivable Traits][debug trait] appendix for more detail about the `Debug` trait and its usage for printing value when debugging.
+For more details about the `Debug` trait and its usage for printing values when debugging, please refer to the [Derivable Traits][debug trait] appendix.
```
Comment (misicnenad) [inline src/ch11-09-printing.md:97]: More natural wording.

File: src/ch11-09-printing.md [hunk 10]
```diff
@@ -84,16 +84,16 @@ It is also possible to implement the `Display` trait for the `Point` struct usin
...
 
-[debug trait]: ./appendix-03-derivable-traits.md#debug-trait-for-printing-and-debugging
\ No newline at end of file
+[debug trait]: ./appendix-03-derivable-traits.md#debug-trait-for-printing-and-debugging
```
Comment (misicnenad) [inline src/ch11-09-printing.md:99]: I'm not sure what the update here is, maybe a different newline character?

File: src/ch11-01-custom-data-structures.md (hunk 5 above)
Comment (TAdev0) [reply src/ch11-01-custom-data-structures.md:24]: yeah just checked the ui doesnt change

File: src/ch11-01-custom-data-structures.md (hunk 6 above)
Comment (TAdev0) [reply src/ch11-01-custom-data-structures.md:115]: its a bit of a repetition but its nice to print it i think

File: src/ch11-01-custom-data-structures.md [hunk 11]
```diff
@@ -177,9 +183,8 @@ The code uses the `insert` and `get` methods to access the values in the
 `Felt252Dict<T>`. To push an element to the top of the stack, the `push`
 function inserts the element in the dict at index `len` and increases the
 `len` field of the stack to keep track of the position of the stack top. To
-remove a value, the `pop` function retrieves the last value at position `len-1`
-and then decreases the value of `len` to update the position of the stack top
-accordingly.
+remove a value, the `pop` function decreases the value of `len` to update the
```
Comment (TAdev0) [inline src/ch11-01-custom-data-structures.md:186]: great!

File: src/ch11-03-smart-pointers.md [hunk 12]
```diff
@@ -10,7 +10,7 @@ The Cairo VM memory is composed by multiple segments that can store data, each i
 
 ## The `Box<T>` Type to Manipulate Pointers
 
-The principal smart pointer type in Cairo is a _box_, denoted as `Box<T>`. Manually defining boxes allow you to store data in a specific memory segment of the Cairo VM called the _boxed_segment_. This segment is dedicated to store all boxed values, and what remains in the execution segment is only a pointer to the boxed segment. Whenever you instantiate a new pointer variable of type `Box<T>`, you append the data of type `T` to the boxed segment.
+The principal smart pointer type in Cairo is a _box_, denoted as `Box<T>`. Manually defining boxes allow you to store data in a specific memory segment of the Cairo VM called _the boxed segment_. This segment is dedicated to store all boxed values, and what remains in the execution segment is only a pointer to the boxed segment. Whenever you instantiate a new pointer variable of type `Box<T>`, you append the data of type `T` to the boxed segment.
```
Comment (TAdev0) [inline src/ch11-03-smart-pointers.md:13]: mmh why did you add ''the'' in italic ?

File: src/ch11-03-smart-pointers.md [hunk 13]
```diff
@@ -35,7 +35,7 @@ Listing {{#ref basic_box}} shows how to use a box to store a value in the boxed
...
+We define the variable `b` to have the value of `Box` that points to the value `5`, which is stored in the boxed segment. This program will print `b = 5`; in this case, we can access the data in the box similar to how we would if this data was simply in the execution memory. Putting a single value in a box isn’t very useful, so you won’t use boxes by themselves in this way very often. Having values like a single `u128` in the execution memory, where they’re stored by default, is more appropriate in the majority of situations. Let’s look at a case where boxes allow us to define types that we wouldn’t be allowed to if we didn’t have boxes.
```
Comment (TAdev0) [inline src/ch11-03-smart-pointers.md:38]: i think its better to keep `a` here

you would say `.. to have the value of a bool`. not ` to have the value of bool`

File: quizzes/ch11-03-smart_pointers.toml [hunk 14]
```diff
@@ -34,6 +34,13 @@ fn main() {
 """
 answer.doesCompile = true
 answer.stdout = "Peter"
+context = """
+The `student1` variable was first instantiated as a smart pointer to an instance of the struct `Student`.
+When we assigned `student1` to a new variable `student2`, we merely copied the _smart pointer_ to the previously created struct,
+so both variables pointed to the same struct in memory.
```
Comment (enitrat) [inline quizzes/ch11-03-smart_pointers.toml:40]:
```suggestion
so both variables referred to the same struct in memory.
```

I like to keep `point` for pointer-related terms

File: quizzes/ch11-03-smart_pointers.toml [hunk 15]
```diff
@@ -34,6 +34,13 @@ fn main() {
...
+Once `student1` was reinstantiated with a new smart pointer to a new `Student` instance, the `student2` variable
+still pointed to the original struct, so printing `student2.name` displayed `"Peter"`.
```
Comment (enitrat) [inline quizzes/ch11-03-smart_pointers.toml:42]:
```suggestion
still referred to the original struct, so printing `student2.name` displayed `"Peter"`.
```

**Review by TAdev0 — COMMENTED** [review]
@misicnenad  thanks a lot for this PR. Very nice improvements overall. I left 1 or 2 minor comments, otherwise everything looks really good!

File: src/ch11-03-smart-pointers.md (hunk 12 above)
Comment (enitrat) [inline src/ch11-03-smart-pointers.md:13]:
```suggestion
The principal smart pointer type in Cairo is a _box_, denoted as `Box<T>`. Manually defining boxes allow you to store data in a specific memory segment of the Cairo VM called the _boxed segment_. This segment is dedicated to store all boxed values, and what remains in the execution segment is only a pointer to the boxed segment. Whenever you instantiate a new pointer variable of type `Box<T>`, you append the data of type `T` to the boxed segment.
```

the emphasis should be on the name of the segment only

File: src/ch11-03-smart-pointers.md (hunk 13 above)
Comment (enitrat) [inline src/ch11-03-smart-pointers.md:38]:
```suggestion
We define the variable `b` to have the value of a `Box` that points to the value `5`, which is stored in the boxed segment. This program will print `b = 5`; in this case, we can access the data in the box similar to how we would if this data was simply in the execution memory. Putting a single value in a box isn’t very useful, so you won’t use boxes by themselves in this way very often. Having values like a single `u128` in the execution memory, where they’re stored by default, is more appropriate in the majority of situations. Let’s look at a case where boxes allow us to define types that we wouldn’t be allowed to if we didn’t have boxes.
```

I think it reads better, not strongly opinionated

File: src/ch11-07-inlining-in-cairo.md [hunk 16]
```diff
@@ -17,15 +18,15 @@ Many of the Cairo corelib functions are inlined. User-defined functions may also
 
 However, inlining can also lead to increased code size. Whenever a function is inlined, the call site contains a copy of the function's Sierra code, potentially leading to duplication of code across the compiled code.
 
-Therefore, inlining should be applied with cautious. Using the `#[inline]` or `#[inline(always)]` indiscriminately will lead to increased compile time. It is particularly useful to inline small functions, ideally with many arguments. This is because inlining large functions will increase the code length of the program, and handling many arguments will increase the number of steps to execute these functions.
+Therefore, inlining should be applied with cautioun. Using `#[inline]` or `#[inline(always)]` indiscriminately will lead to increased compile time. It is particularly useful to inline small functions, ideally with many arguments. This is because inlining large functions will increase the code length of the program, and handling many arguments will increase the number of steps to execute these functions.
```
Comment (enitrat) [inline src/ch11-07-inlining-in-cairo.md:21]:
```suggestion
Therefore, inlining should be applied with caution. Using `#[inline]` or `#[inline(always)]` indiscriminately will lead to increased compile time. It is particularly useful to inline small functions, ideally with many arguments. This is because inlining large functions will increase the code length of the program, and handling many arguments will increase the number of steps to execute these functions.
```

File: src/ch11-07-inlining-in-cairo.md [hunk 17]
```diff
@@ -132,15 +137,16 @@ We can now decompose how these instructions are executed to understand what this
 - `ret`: jumps back to the line after `call rel 3`, so we go to line 2.
 - `ret`: last instruction executed as there is no more `call` instruction where to jump right after. This is the actual return instruction of the Cairo `main` function.
 
-To summary: 
+To summary:
```
Comment (enitrat) [inline src/ch11-07-inlining-in-cairo.md:140]:
```suggestion
To summarize:
```

File: src/ch11-09-printing.md [hunk 18]
```diff
@@ -6,10 +6,10 @@ When writing a program, it is quite common to print some data to the console, ei
 
 Cairo provides two macros to print standard data types:
 
-- `println!` which prints on a new line,
+- `println!` which prints on a new line.
 - `print!` with inline printing.
```
Comment (enitrat) [inline src/ch11-09-printing.md:10]:
```suggestion
- `println!` which prints on a new line
- `print!` with inline printing
```

no full stop when they're not full sentences

**Review by enitrat — APPROVED** [review]
a pleasure to review :)

File: listings/ch11-advanced-features/no_listing_12_dict_struct_member/src/lib.cairo [hunk 19]
```diff
@@ -43,7 +43,7 @@ impl UserDatabaseDestruct<T, +Drop<T>, +Felt252DictValue<T>> of Destruct<UserDat
 
 // ANCHOR: main
 fn main() {
-    let mut db = UserDatabaseTrait::new();
+    let mut db = UserDatabaseTrait::<u64>::new();
```
[the remark this answers is not in the document]
Comment (misicnenad) [reply listings/ch11-advanced-features/no_listing_12_dict_struct_member/src/lib.cairo:46]: I could've sworn I had a compiler error here, this is weird :open_mouth:
Maybe I had an unrelated error

File: src/ch11-03-smart-pointers.md (hunk 12 above)
Comment (misicnenad) [reply src/ch11-03-smart-pointers.md:13]: seemed like it's part of the name, will move it out

File: src/ch11-07-inlining-in-cairo.md (hunk 16 above)
Comment (misicnenad) [reply src/ch11-07-inlining-in-cairo.md:21]: *sigh* missed the "u"

**Comment by misicnenad** [comment]
@enitrat I *think* I applied all of the suggestions

**Review by TAdev0 — APPROVED** [review]
@misicnenad  just checked your last commits
lgtm!
